sklearn/cross_decomposition/_pls.py

```diff
@@ -239,10 +239,17 @@ def fit(self, X, Y):
                 Yk_mask = np.all(np.abs(Yk) < 10 * Y_eps, axis=0)
                 Yk[:, Yk_mask] = 0.0
 
-                x_weights, y_weights, n_iter_ = \
-                    _get_first_singular_vectors_power_method(
-                        Xk, Yk, mode=self.mode, max_iter=self.max_iter,
-                        tol=self.tol, norm_y_weights=norm_y_weights)
+                try:
```

Comment:
I was more thinking about doing the try/except on the offending line:

`y_score = next(col for col in Y.T if np.any(np.abs(col) > eps))`

Is there a way we can move it there while keeping the logic here reasonably sane?
My concern is that it's not obvious which line can raise a StopIteration if we leave the try/except here.

Reply:
If we catch it at `y_score`, we still need to raise another exception to tell the caller to break out of the loop.

I added [`3d4b6d2` (#19922)](https://github.com/scikit-learn/scikit-learn/pull/19922/commits/3d4b6d25fa8ddba2d093b496146746309bc7a27f) to make the control flow a little more obvious.